smartcontract/service/native/param.go

```diff
@@ -0,0 +1,161 @@
+	scommon "github.com/ontio/ontology/core/store/common"
+	ctypes "github.com/ontio/ontology/core/types"
+	"github.com/ontio/ontology/errors"
+	"github.com/ontio/ontology/smartcontract/service/native/states"
+	"sync"
+)
+
+type ParamCache struct {
+	lock   sync.RWMutex
+	Params map[string]string
+}
+
+var GLOBAL_PARAM = map[string]string{
+	"init-key1": "init-value1",
+	"init-key2": "init-value2",
+	"init-key3": "init-value3",
+	"init-key4": "init-value4",
+}
+
+var paramCache *ParamCache
+var admin *states.Admin
+
+func init() {
+	Contracts[genesis.ParamContractAddress] = RegisterParamContract
+	paramCache = new(ParamCache)
+	paramCache.Params = make(map[string]string)
+}
+
+func ParamInit(native *NativeService) error {
+	paramCache = new(ParamCache)
+	paramCache.Params = make(map[string]string)
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	for k, v := range GLOBAL_PARAM {
+		native.CloneCache.Add(scommon.ST_STORAGE, getParamKey(contract, k), getParamStorageItem(v))
+		paramCache.Params[k] = v
+	}
+	admin = new(states.Admin)
+	admin.Address = ctypes.AddressFromPubKey(account.GetBookkeepers()[0])
+	native.CloneCache.Add(scommon.ST_STORAGE, getAdminKey(contract), getAdminStorageItem(admin))
+	return nil
+}
+
+func TransferAdmin(native *NativeService) error {
+	destinationAdmin := new(states.Admin)
+	if err := destinationAdmin.Deserialize(bytes.NewBuffer(native.Input)); err != nil {
+		return errors.NewErr("[Transfer Admin]Deserialize Admins failed!")
+	}
+	if !native.ContextRef.CheckWitness(destinationAdmin.Address) {
+		return errors.NewErr("[Transfer Admin]Authentication failed!")
+	}
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	initAdmin(native, contract)
+	transferAdmin, err := getStorageAdmin(native, getTransferAdminKey(contract, admin.Address, destinationAdmin.Address))
+	if err != nil || transferAdmin.Address != destinationAdmin.Address {
+		return errors.NewDetailErr(err, errors.ErrNoCode,
+			"[Transfer Admin] Destination account hasn't been approved!")
+	}
+	// delete transfer admin item
+	native.CloneCache.Delete(scommon.ST_STORAGE, getTransferAdminKey(contract, admin.Address, destinationAdmin.Address))
+	// modify admin in database
+	native.CloneCache.Add(scommon.ST_STORAGE, getAdminKey(contract), getAdminStorageItem(destinationAdmin))
+
+	admin = destinationAdmin
+	return nil
+}
+
+func ApproveAdmin(native *NativeService) error {
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	initAdmin(native, contract)
+	if !native.ContextRef.CheckWitness(admin.Address) {
+		return errors.NewErr("[Approve Admin]Authentication failed!")
+	}
+	destinationAdmin := new(states.Admin)
+	if err := destinationAdmin.Deserialize(bytes.NewBuffer(native.Input)); err != nil {
+		return errors.NewErr("[Approve Admin]Deserialize Admins failed!")
+	}
+	native.CloneCache.Add(scommon.ST_STORAGE, getTransferAdminKey(contract, admin.Address, destinationAdmin.Address),
```

Comment:
approve key should unique, or else repeated approve transaction may cause problem

Reply:
the key is constructed by origin admin address and new admin address, if approve to more than one account and once a account invoke transferAdmin, the admin address of other account will change.

Reply:
consider those case: 1, unique key can support cancel operation but this implementation cannot, if A has approved B, A can approve to himself  to cancel the previous approve. 2. if A approve B, C, B became admin, and some time later approve back to A.  A became admin again. but C still has the right, which maybe  against A's expectation.

Reply:
consider those case: 1, unique key can support cancel operation but this implementation cannot, if A has approved B, A can approve to himself to cancel the previous approve. 2. if A approve B, C, B became admin, and some time later approve back to A. A became admin again. but C still has the right, which maybe against A's expectation.

Reply:
OK, I delete the other key-value in database while a account accept admin.